feat: Checking finalized sizes + a test of two folding verifiers

codygunton opened the pull request:
Original goal: add a test so we can get the size of a circuit containing two folding verifiers.

Detour: (Ultra and Mega flavors only) getting the size after adding nonzero gates and finalizing a circuit used to require adding a printout to the Honk proving key constructor, which would either be noise in many tests, or which would then have to be removed. Rearranging the ensure_nonzero function, we can accommodate the use case in tests where we want to print out the number of gates after finalization. To do this, just call finalize and and print the number of gates before passing the circuit to the prover or proving key constructor. You get a warning, but that's appropriate.

Caution: I left in a code path that finalizes but does add extra gates because this was in use in many places. If those are all valid uses, then we only really want the gate-adding version in one place, so I made the default not add nonzero gates.

```
Console is in 'commands' mode, prefix expressions with '?'.
Launching: /mnt/user-data/cody/aztec-packages/barretenberg/cpp/build-debug/bin/stdlib_protogalaxy_verifier_tests --gtest_color=no --gtest_filter=ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest --gtest_also_run_disabled_tests --gtest_break_on_failure
Launched process 2323108
Running main() from /mnt/user-data/cody/aztec-packages/barretenberg/cpp/build-debug/_deps/gtest-src/googletest/src/gtest_main.cc
Note: Google Test filter = ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ProtogalaxyRecursiveTests/0, where TypeParam = bb::MegaRecursiveFlavor_<bb::MegaCircuitBuilder_<bb::field<bb::Bn254FrParams> > >
[ RUN      ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
Folding Recursive Verifier: num gates unfinalized = 18085
Folding Recursive Verifier: num gates finalized = 20211
WARNING: Redundant call to finalize_circuit(). Is this intentional?
Dyadic size of verifier circuit: 32768
[       OK ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest (6768 ms)
[----------] 1 test from ProtogalaxyRecursiveTests/0 (6768 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (6768 ms total)
[  PASSED  ] 1 test.
Process exited with code 0.
```

codygunton committed:
Add test to check size of two folding verifiers. Result:

```
Running main() from /mnt/user-data/cody/aztec-packages/barretenberg/cpp/build/_deps/gtest-src/googletest/src/gtest_main.cc
Note: Google Test filter = ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ProtogalaxyRecursiveTests/0, where TypeParam = bb::MegaRecursiveFlavor_<bb::MegaCircuitBuilder_<bb::field<bb::Bn254FrParams> > >
[ RUN      ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
Folding Recursive Verifier: num gates unfinalized = 37272
Folding Recursive Verifier: num gates finalized = 44041
WARNING: Redundant call to finalize_circuit(). Is this intentional?
Dyadic size of verifier circuit: 65536
[       OK ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest (1854 ms)
[----------] 1 test from ProtogalaxyRecursiveTests/0 (1854 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1854 ms total)
[  PASSED  ] 1 test.
```

codygunton committed:
Fix typo in error

codygunton committed:
Fix typo

codygunton committed:
hacky fix

codygunton committed:
Fix `ensure` count

codygunton committed:
Revert "hacky fix"

This reverts commit 6998193981731301a598c43d3224e29a251373ff.

codygunton committed:
Merge remote-tracking branch 'origin/master' into cg/two-folding-verifiers-test

codygunton commented on barretenberg/cpp/src/barretenberg/srs/factories/mem_bn254_crs_factory.cpp:
```diff
@@ -60,7 +60,7 @@ MemBn254CrsFactory::MemBn254CrsFactory(std::vector<g1::affine_element> const& po
 std::shared_ptr<bb::srs::factories::ProverCrs<curve::BN254>> MemBn254CrsFactory::get_prover_crs(size_t degree)
 {
     if (prover_crs_->get_monomial_size() < degree) {
-        throw_or_abort(format("prover trying to get too many points in MemGrumpkinCrsFactory! ",
```
Typo in error message.

lucasxia01 replied:
whoops

codygunton commented on barretenberg/cpp/src/barretenberg/stdlib/protogalaxy_verifier/protogalaxy_recursive_verifier.test.cpp:
```diff
@@ -405,6 +410,11 @@ TYPED_TEST(ProtogalaxyRecursiveTests, RecursiveFoldingTest)
     TestFixture::test_recursive_folding();
 }
 
+TYPED_TEST(ProtogalaxyRecursiveTests, RecursiveFoldingTwiceTest)
```
The new test from refactoring the other test.

codygunton commented on barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/ultra_circuit_builder.hpp:
```diff
@@ -421,7 +421,7 @@ class UltraCircuitBuilder_ : public CircuitBuilderBase<typename Arithmetization_
     }
 
-    void finalize_circuit();
+    void finalize_circuit(const bool ensure_nonzero = false);
```
I wanted to preserve the ability to finalize without ensuring nonzero because this is how the function is used in most places (e.g. in the SMT solver). I should raise this issue internally to make sure people know this is happening.

lucasxia01 replied:
I feel like this makes things more confusing. Like we're going to have unfinalized gate count, sorta finalized gate count without ensure_nonzero gates, and also a finalized gate count...

codygunton replied:
Yeah I'd rather not add this but I also don't want to change underlying assumptions--literally only one case DOES ensure gates are nonzero...

lucasxia01 commented on barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/ultra_circuit_builder.cpp:
```diff
@@ -41,6 +42,9 @@ template <typename Arithmetization> void UltraCircuitBuilder_<Arithmetization>::
      */
     if (!circuit_finalized) {
+        if (ensure_nonzero) {
+            add_gates_to_ensure_all_polys_are_non_zero();
```
confusing that this wasn't renamed to add_ultra_gates_to_ensure_all_polys_are_non_zero

codygunton replied:
It's a member function of UltraCircuitBuilder so it feels redundant.

lucasxia01 commented on barretenberg/cpp/src/barretenberg/stdlib/protogalaxy_verifier/protogalaxy_recursive_verifier.test.cpp:
```diff
@@ -226,7 +227,11 @@ template <typename RecursiveFlavor> class ProtogalaxyRecursiveTests : public tes
         if constexpr (!IsSimulator<OuterBuilder>) {
+            // inefficiently check finalized size
+            folding_circuit.finalize_circuit();
+            info("Folding Recursive Verifier: num gates finalized = ", folding_circuit.num_gates);
```
so this is now printing out finalized gates without the ensure nonzero gates? But in practice, our circuit is actually going to include these ensure_nonzero gates so I don't understand why we wouldn't want them.

codygunton replied:
Yes--before the PR that was the behavior of this test, so I'm just preserving the test. Whether it _should_ be that behavior... I'll make an issue.

codygunton replied:
https://github.com/AztecProtocol/barretenberg/issues/1100

lucasxia01 replied:
could link this to code but also no need if you just want to merge

lucasxia01 reviewed (commented):
FoldingVerifier test looks great. I'm somewhat unclear why we want this finalized-but-without-ensure-nonzero-gates gate count because it seems to make things more confusing. It seems like we could already add finalize_circuit calls to tests to print out the finalized gate count like you said without this finalized-but-without-ensure-nonzero-gates thing.

codygunton committed:
Make test more idiomatic so it doesn't re-oink the accumulator. Result:

Console is in 'commands' mode, prefix expressions with '?'.
Launching: /mnt/user-data/cody/aztec-packages/barretenberg/cpp/build-debug/bin/stdlib_protogalaxy_verifier_tests --gtest_color=no --gtest_filter=ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest --gtest_also_run_disabled_tests --gtest_break_on_failure
Launched process 2317410
Running main() from /mnt/user-data/cody/aztec-packages/barretenberg/cpp/build-debug/_deps/gtest-src/googletest/src/gtest_main.cc
Note: Google Test filter = ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ProtogalaxyRecursiveTests/0, where TypeParam = bb::MegaRecursiveFlavor_<bb::MegaCircuitBuilder_<bb::field<bb::Bn254FrParams> > >
[ RUN      ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
Folding Recursive Verifier: num gates unfinalized = 18085
Folding Recursive Verifier: num gates finalized = 20194
WARNING: Redundant call to finalize_circuit(). Is this intentional?
Dyadic size of verifier circuit: 32768
[       OK ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest (13082 ms)
[----------] 1 test from ProtogalaxyRecursiveTests/0 (13082 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (13082 ms total)
[  PASSED  ] 1 test.
Process exited with code 0.

codygunton committed:
Merge remote-tracking branch 'origin/master' into cg/two-folding-verifiers-test

codygunton committed:
Fix typo to count gates with ensuring nonzero. Result:

```
Console is in 'commands' mode, prefix expressions with '?'.
Launching: /mnt/user-data/cody/aztec-packages/barretenberg/cpp/build-debug/bin/stdlib_protogalaxy_verifier_tests --gtest_color=no --gtest_filter=ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest --gtest_also_run_disabled_tests --gtest_break_on_failure
Launched process 2323108
Running main() from /mnt/user-data/cody/aztec-packages/barretenberg/cpp/build-debug/_deps/gtest-src/googletest/src/gtest_main.cc
Note: Google Test filter = ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ProtogalaxyRecursiveTests/0, where TypeParam = bb::MegaRecursiveFlavor_<bb::MegaCircuitBuilder_<bb::field<bb::Bn254FrParams> > >
[ RUN      ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest
Folding Recursive Verifier: num gates unfinalized = 18085
Folding Recursive Verifier: num gates finalized = 20211
WARNING: Redundant call to finalize_circuit(). Is this intentional?
Dyadic size of verifier circuit: 32768
[       OK ] ProtogalaxyRecursiveTests/0.RecursiveFoldingTwiceTest (6768 ms)
[----------] 1 test from ProtogalaxyRecursiveTests/0 (6768 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (6768 ms total)
[  PASSED  ] 1 test.
Process exited with code 0.
```

lucasxia01 commented on barretenberg/cpp/src/barretenberg/stdlib/protogalaxy_verifier/protogalaxy_recursive_verifier.test.cpp:
```diff
@@ -204,14 +204,30 @@ template <typename RecursiveFlavor> class ProtogalaxyRecursiveTests : public tes
+        std::shared_ptr<RecursiveDeciderVerificationKey> accumulator;
+        for (size_t idx = 0; idx < num_verifiers; idx++) {
+            accumulator = verifier.verify_folding_proof(stdlib_proof);
+            if (idx < num_verifiers - 1) { // else the transcript is null in the test below
```
wait, how was there no error from the transcript being null before (when I first reviewed)?

lucasxia01 reviewed (approved):
after some discussion, LGTM, but only if this confusing code gets fixed soon
